src/compiler/program.ts

```diff
@@ -453,6 +453,66 @@ module ts {
             }
         }
 
+        function computecommonSourceDirectory(sourceFiles: SourceFile[]): string {
+            let commonPathComponents: string[];
+            let currentDirectory = host.getCurrentDirectory();
+            forEach(files, sourceFile => {
+                // Each file contributes into common source file path
+                if (isDeclarationFile(sourceFile)) {
+                    return;
+                }
+
+                let sourcePathComponents = getNormalizedPathComponents(sourceFile.fileName, currentDirectory);
+                sourcePathComponents.pop(); // FileName is not part of directory
+                if (commonPathComponents) {

```

Comment:
I'm saying instead of having a big `if` statement body, just check perform the `else` clause when applicable